csrc/scheduler/normalization_inner_outer.cpp

```diff
@@ -182,11 +182,252 @@ bool InnerOuterPersistentKernelScheduler::canScheduleCompileTime(
+}
+
+// Sorts projectable_buffer_inputs based on the latency from smallest to
+// largest. The latency is calculated based on the number of uses, the buffer
+// data type size, and the pre-defined schedule structures. Assume each
+// global memory load has latency 100, each shared memory load has latency 10,
+// latency is proportional to the data type size. There is no write from
+// register to shared memory, so this type of latency is not considered. The
+// magic number 100 and 10 are not critical for our current major use case, e.g.
+// layer norm backward, and rms norm backward. The outer broadcast tensor (the
+// weight tensor) is almost always moved to shared memory before other buffers
+// unless the outer dimension is very small, e.g. less than 1 sm count or other
+// buffers are used 10 times more than the outer broadcast tensor.
+std::vector<TensorView*> sortProjectableBufferInputs(
+    Fusion* fusion,
+    const std::vector<TensorView*>& persistent_buffers_original,
+    const std::vector<TensorView*>& projectable_buffer_inputs,
+    const std::vector<TensorView*>& outer_broadcast_tvs,
+    const std::vector<TensorView*>& reduction_tvs,
+    SchedulerRuntimeInfo& runtime_info) {
+  const int64_t n_buffer_inputs = (int64_t)projectable_buffer_inputs.size();
+
+  // Set the number of uses for each projected buffer inputs.
+  // Includes direct use and the use of its corresponding original persistent
+  // buffer. When project to inputs, we are replacing the use of the original
+  // persistent buffers except for the one goes to reduction.
+  std::vector<int> n_proj_buffer_uses(n_buffer_inputs, 0);
+  for (auto idx = 0; idx < n_buffer_inputs; idx++) {
+    auto input_tv = projectable_buffer_inputs[idx];
+    n_proj_buffer_uses[idx] = (int)input_tv->uses().size();
+    for (auto buffer : persistent_buffers_original) {
+      if (DependencyCheck::isDependencyOf(input_tv, buffer)) {
+        n_proj_buffer_uses[idx] += (int)buffer->uses().size() - 1;
+      }
+    }
+  }
+
+  // assume the fusion is scheduled with a two-step approach, the first step
+  // finishes the inner reduction and a partial outer reduction. The second
+  // step finishes the outer reduction. In the first step, the outer dimension
+  // is parallelized by BIDy, each thread loops over the outer dimension
+  // for n_outer_loop = ceilDiv(outer_dim_size, gdimy) times. gdimy is a
+  // heuristic para, here we assume gdimy = sm_count. For outer broadcast tv,
+  // e.g. weight tensor, it is reused in every outer loop, when calculate the
+  // latency of global memory load, 1 is used instead of n_outer_loop. For other
+  // tensors, each loop is loading a different row, so n_outer_loop is used.
+  std::vector<int> buffer_latency(n_buffer_inputs, 0);
+  int sm_count =
+      (int)at::cuda::getCurrentDeviceProperties()->multiProcessorCount;
+  int n_outer_loop = [&sm_count, &reduction_tvs, &runtime_info]() -> int {
+    for (auto buffer : reduction_tvs) {
```

Comment:
Are all of the outer reduction tensors in `reduction_tvs` guaranteed to have the same shape? Asking because it seems there can be other tensors that have broadcast domains that correspond to the reduction domains in these `reduction_tvs`.

Reply:
yes, `compile time can schedule` check ensures they have the same shape.

Reply:
Can you please add a comment?

Reply:
added.
```
      // Use the first outer reduction tensor to infer the outer dim size.
      // Compile time check ensures they have the same iter and reduction axes,
      // connected with inner reduction tv. Similar to the first inner reduction
      // tv is used to derive the reduction properties. Here the first outer
      // reduction tv is used to derive the outer dim size.
```